### PR TITLE
Add `letter-spacing` property to RCSS

### PR DESCRIPTION
**Pull request by igorsegallafa**
Related to https://github.com/mikke89/RmlUi/issues/426

image

image

### CODE REVIEW

File: Source/Core/ElementStyle.cpp [hunk 1]
```diff
@@ -784,6 +784,10 @@ PropertyIdSet ElementStyle::ComputeValues(Style::ComputedValues& values, const S
 			// (font-size computed above)
 			dirty_font_face_handle = true;
 			break;
+		case PropertyId::LetterSpacing:
+			values.letter_spacing(ComputeLength(p, font_size, document_font_size, dp_ratio, vp_dimensions));
```
Comment (mikke89) [inline Source/Core/ElementStyle.cpp:788]:
```suggestion
			values.letter_spacing(p->unit == Property::KEYWORD ? 0.f : ComputeLength(p, font_size, document_font_size, dp_ratio, vp_dimensions));
```
We only have a single keyword which computes to zero, otherwise resolve the length.

File: Source/Core/ElementUtilities.cpp [hunk 2]
```diff
@@ -154,13 +154,13 @@ float ElementUtilities::GetDensityIndependentPixelRatio(Element * element)
 }
 
 // Returns the width of a string rendered within the context of the given element.
-int ElementUtilities::GetStringWidth(Element* element, const String& string, Character prior_character)
+int ElementUtilities::GetStringWidth(Element* element, const String& string, float letter_spacing, Character prior_character)
```
Comment (mikke89) [inline Source/Core/ElementUtilities.cpp:157]: Rather than taking this as a parameter here, we could retrieve it from the `element`'s computed values. This will also simplify many of the calls to this function.

File: Source/Core/ElementText.cpp [hunk 3]
```diff
@@ -369,6 +373,8 @@ void ElementText::OnPropertyChange(const PropertyIdSet& changed_properties)
 		font_effects_handle = 0;
 		font_effects_dirty = true;
 		font_handle_version = 0;
+
+		letter_spacing = computed.letter_spacing();
```
Comment (mikke89) [inline Source/Core/ElementText.cpp:377]: This value is already stored in the element's computed values, so we don't need to store it as a member of this class. Instead, retrieve it from our computed values as needed.

File: Include/RmlUi/Core/FontEngineInterface.h [hunk 4]
```diff
@@ -93,7 +93,7 @@ class RMLUICORE_API FontEngineInterface
 	/// @param[in] string The string to measure.
 	/// @param[in] prior_character The optionally-specified character that immediately precedes the string. This may have an impact on the string width due to kerning.
 	/// @return The width, in pixels, this string will occupy if rendered with this handle.
-	virtual int GetStringWidth(FontFaceHandle handle, const String& string, Character prior_character = Character::Null);
+	virtual int GetStringWidth(FontFaceHandle handle, const String& string, float letter_spacing, Character prior_character = Character::Null);
```
Comment (mikke89) [inline Include/RmlUi/Core/FontEngineInterface.h:96]: Add parameter description in comment.

File: Include/RmlUi/Core/FontEngineInterface.h [hunk 5]
```diff
@@ -105,7 +105,7 @@ class RMLUICORE_API FontEngineInterface
 	/// @param[out] geometry An array of geometries to generate the geometry into.
 	/// @return The width, in pixels, of the string geometry.
 	virtual int GenerateString(FontFaceHandle face_handle, FontEffectsHandle font_effects_handle, const String& string, const Vector2f& position,
-		const Colourb& colour, float opacity, GeometryList& geometry);
+		const Colourb& colour, float opacity, float letter_spacing, GeometryList& geometry);
```
Comment (mikke89) [inline Include/RmlUi/Core/FontEngineInterface.h:108]: Add parameter description in comment.

**Review by mikke89 — COMMENTED** [review]
You've figured your way around the code base very well. Most of my comments are relatively minor fixups.

Could you add a VisualTest as well? Something that shows the output with different letter-spacing. Maybe something with line-breaks or other edge-cases too?
To get started, you could base the test on another test, such as `Tests/Data/VisualTests/word_break.rml`.

Just an FYI for the future, you can change the base branch of PRs on GitHub itself. There shouldn't be a need to open new PRs, no worries though.

**Comment by igorsegallafa** [comment]
Good thoughts, thank you! 
ll take some time to implement the tests right now.

**Comment by mikke89** [comment]
Hey, I decided to finish this one up myself, including a visual test, and some efficiency considerations. Thanks for the pull request!
